WordPressKit/AccountServiceRemoteREST.m

```diff
@@ -127,31 +127,54 @@ - (void)isPasswordlessAccount:(NSString *)identifier success:(void (^)(BOOL pass
+        if (!success) {
+            return;
+        }
 
-              // If the email address is not available (has already been used)
-              // the endpoint will reply with a 200 status code and an JSON
-              // object describing an error.
-              // The error is that the queried email address is not available,
-              // which is our failure case. Test the error response for the
-              // "taken" reason to confirm the email address exists.
-              BOOL available = NO;
-              if ([responseObject isKindOfClass:[NSDictionary class]]) {
-                  NSDictionary *dict = (NSDictionary *)responseObject;
-                  available = [[dict numberForKey:@"available"] boolValue];
-              }
-              success(available);
+        if ([responseObject isKindOfClass:[NSDictionary class]]) {
+            NSString *error = [responseObject objectForKey:@"error"];
+            NSString *message = [responseObject objectForKey:@"message"];
+            
+            if ([error isEqualToString:errorEmailAddressInvalid]) {
+                NSError* error = [[NSError alloc] initWithDomain:AccountServiceRemoteErrorDomain
+                                                            code:AccountServiceRemoteEmailAddressInvalid
+                                                        userInfo:@{
+                                                            @"response": responseObject,
+                                                            NSLocalizedDescriptionKey: message,
+                                                        }];
+                failure(error);
+                return;
+            } else if ([error isEqualToString:errorEmailAddressTaken]) {
+                NSError* error = [[NSError alloc] initWithDomain:AccountServiceRemoteErrorDomain
+                                                            code:AccountServiceRemoteEmailAddressTaken
+                                                        userInfo:@{
+                                                            @"response": responseObject,
+                                                            NSLocalizedDescriptionKey: message,
+                                                        }];
+                failure(error);
+                return;
+            }
+            
+            BOOL available = [[responseObject numberForKey:@"available"] boolValue];
+            success(available);
+        } else {
+            NSError* error = [[NSError alloc] initWithDomain:AccountServiceRemoteErrorDomain
+                                                        code:AccountServiceRemoteCantReadServerResponse
+                                                    userInfo:@{@"response": responseObject}];
 
-          } failure:^(NSError *error, NSHTTPURLResponse *httpResponse) {
-              if (failure) {
-                  failure(error);
-              }
-          }];
+            failure(error);
```

Comment:
I've noticed there's no safeguard for `failure = nil` scenarios. Mentioning this only because there is one below!

Reply:
Addressed.